### ISSUE TITLE
BinActive take input mean into consideration?

### ISSUE THREAD

**Issue by zhaohui-yang** [jiecaoyu/XNOR-Net-PyTorch#56]
Thanks for sharing this perfect project!

In XNOR-net, the author calculate mean map and multiply it after convolution, but I haven't found any code about this in BinActive function. I wonder if you merge it with other part? or somewhere? Thank you !

```python
class BinActive(torch.autograd.Function):
    #Binarize the input activations and calculate the mean across channel dimension.

    def forward(self, input):
        self.save_for_backward(input)
        size = input.size()
        input = input.sign()
        return input

    def backward(self, grad_output):
        input, = self.saved_tensors
        grad_input = grad_output.clone()
        grad_input[input.ge(1)] = 0
        grad_input[input.le(-1)] = 0
        return grad_input
```

**Comment by zhaohui-yang** [comment]
By the way, I just implement adding input alpha into your mnist example, and achieved 99.29 accuracy which is higher than not adding this.

**Comment by jiecaoyu** [comment]
@zhaohui-yang Hi, I include the scaling factors for the weights (alpha) but not the scaling factors for the inputs (beta). Yes, including the scaling factors for the inputs can help increase the accuracy. But it will slow download training process. Considering the accuracy improvement is limited, I do not include the scaling factors for the inputs.

**Comment by zhaohui-yang** [comment]
@jiecaoyu  Thank you, it really helps!

**Comment by zhaohui-yang** [comment]
@jiecaoyu  Have you ever tried HORQ-net on Imagenet? As you said, "the accuracy improvement is limited". In HORQ-net, they tried to approximate input using residual, but only released MNIST and CIFAR results. I'm not sure whether more accurate input would effert the result a lot? or as you said, improvement is limited?

**Comment by PeiyanFlying** [comment]
Hi, I think the math function y=sign(x) is when x<0 y=-1, when x=0 y=0, when x>0 y=1. So in this code is grad_input[input.ge(1)] = 0; grad_input[input.le(-1)] = 0? Not grad_input[input.gt(1)] = 0
        grad_input[input.lt(-1)] = 0?

**Comment by jiecaoyu** [comment]
Yes. But because of the weight decay, these two implementations will have similar results.
